PARAMETERS.md
# Algorithm parameters

To create a customized configuration parameter file use the included example file [example_options_file.txt](example/example_options_file.txt) as a base and modify it. Each mandatory field and its possible values are described below:

| Parameter | Values | Description |
| :---      | :----: | :---        |
| `VERBOSITY` | 0,..,4 | Verbosity level, in increasing order |
| `MAXITER` | 1,... | Maximum number of iterations normalized by the number of scenarios, i.e. `MAXITER`=100 sets a maximum of 100*N* coordinate descent iterations |
| `MAXTIME` | (0, +Inf) | Maximum execution time in seconds |
| `DSHARE_MIDITER`<sup>1</sup> | (0, 1) | Medium point for dynamic change in resource allocation policy for processors in terms of the total number of iterations |
| `DSHARE_START`<sup>1</sup> | (0, 1) | Proportion of processors allocated to solve dual subproblems in the coordinate-descent first iteration |
| `DSHARE_END`<sup>1</sup> | (0, 1) | Proportion of processors allocated to solve dual subproblems at the coordinate-descent last iteration, `MAXITER`*N* |
| `NONSCEN_PERIOD` | (0, 1) | Periodicity to solve subproblem (6), e.g. `NONSCEN_PERIOD`=0.5 would cause that subproblem (6) is solved every two subproblems (7) |
| `NONSCEN_CENTER`<sup>2</sup> | 10, 11 | See *u*<sub>0</sub> in eq. (8). 10: fixed *u*<sub>0</sub> at the average of the first-stage solutions during the first evaluation for each scenario, 11: use incumbent first stage solution as *u*<sub>0</sub> |
| `DUAL_STEPSIZE` | 20, 21, 22 | 20: constant, 21: diminishing, 22: Polyak |
| `DUAL_STEPSIZE_P` | (0, +Inf) | *p*<sub>0</sub> in eq. (14) |
| `DUAL_STEPSIZE_R` | (0, +Inf) | *r* in eq. (14) |
| `DUAL_STEPSIZE_Q` | (0.5, 1] | *q* in eq. (14) |
| `DUAL_STEPSIZE_XI` | (0, 1) | &xi; in eq. (14), expressed in relative terms with respect to *LB*<sub>*k*</sub> |
| `DUAL_STEPSIZE_SGM` | (0, +Inf) | &sigma; in eq. (14) |
| `DUAL_GNORM_FILTER` | (0, 1) | Allows to put a filter in the updating of *g*<sub>*k*</sub> as new information arrives from subproblems. `DUAL_GNORM_FILTER`=0 implies no filter, `DUAL_GNORM_FILTER`=1 implies that *g*<sub>*k*</sub> is not updated. |
| `PRIMAL_RECOVERY` | 30,..,33 | 30: FIFO, 31: Random, 32: LIFO, 33: IS |
| `PRIMAL_IS_SSIZE` | (0, 1) | Relative size of subsample to form average when using IS primal recovery |
| `INIT_TYPE` | 40, 41 | 40: run initialization using LP relaxations, 41: run initialization using period relaxations |
| `DUAL_GAP_TOL` | (0, 1) | Duality gap tolerance for termination; `DUAL_GAP_TOL`=0.01 would terminate the algorithm once the duality gap reaches 1% |
| `SMOOTHING_PARAM`<sup>2</sup> | (0, +Inf) | &mu; in eq. (8) |
| `PROJ_ROUND_MIDP` | (0, 1) | Conservativeness of rounding operations. Variables with non-integer variables will tend to be rounded to 1 if they are `PROJ_ROUND_MIDP` or more. Used during initialization with period relaxation and by IS primal recovery. |
| `PERIOD_AGG_POL` | 50, 51, 52 | Policy for aggregating values of variables appearing in constraints in multiple periods when performing initialization with period relaxation. 50: use minimum value, 51: use average, 52: use maximum value |
| `SCEN_LP_ALG` | p,d,b | Algorithm to solve the root LP relaxation of the scenario subproblem. p: primal simplex, d: dual simplex, b: barrier |
| `PERIOD_LP_ALG` | p,d,b | Algorithm to solve the root LP relaxation of the period subproblem (initialization). p: primal simplex, d: dual simplex, b: barrier |

<sup>1</sup> The proportion of the processors dedicated to dual iterations follows an hyperbolic tangent with *DualShare*=`DSHARE_START` at iteration 1, *DualShare=*`DSHARE_MIDITER` at iteration `MAXITER`*N*/2, and *DualShare*=`DSHARE_END` at iteration `MAXITER`*N*.

<sup>2</sup> For performance of the method and accuracy of the dual function smooth approximation, we recommend setting `NONSCEN_CENTER`=11 and `SMOOTHING_PARAM` to a small value for your problem, e.g. `SMOOTHING_PARAM`=0.1.

The implementation of AsyncLSD, additionally, allows to pass any configuration parameter to Xpress using their code numbers (see `Xpress.h`). These parameters are optional and are declared in the following sections of the configuration file:
	
| Section | Parameters |
| :---    | :---       |
| `XPRESS_SCEN_INT` | Integer parameters for scenario subproblems |
| `XPRESS_SCEN_DBL` | Double parameters for scenario subproblems |
| `XPRESS_PERIOD_INT` | Integer parameters for period-decomposition subproblems and first-stage subproblem |
| `XPRESS_PERIOD_DBL` | Double parameters for period decomposition subproblems and first stage subproblem |


example/example_options_file.txt
# Prototype options file:
# Fixed column positions: names 5 and values 25
# Order of sections must be respected.
# Order of params within GENERAL section must also be respected.
# All parameters in GENERAL section must be input,
# even if they are not necessary.
GENERAL
    VERBOSITY           2         # VERB_COORDINATOR_FUNCS
    MAXITER             200
    MAXTIME             1800
    DSHARE_MIDITER      0.5
    DSHARE_START        0.5
    DSHARE_END          0.5
    NONSCEN_PERIOD      0.1
    NONSCEN_CENTER      11        # use incumbent as center
    DUAL_STEPSIZE       22        # DUAL_STEPSIZE_POLYAK: p/(1+r*k)^q*min(xi*f(x), fup - f(x))/max(sigma, |g|^2_2)
    DUAL_STEPSIZE_P     .5        # 0.5 at the beggining
    DUAL_STEPSIZE_R     0.02      # 0.25 at 50 passes over data
    DUAL_STEPSIZE_Q     1.0       # decreasing at 1/k
    DUAL_STEPSIZE_XI    0.025
    DUAL_STEPSIZE_SGM   1E-8
    DUAL_GNORM_FILTER   0.1
    PRIMAL_RECOVERY     33        # PRIMAL_RECOVERY_IS
    PRIMAL_IS_SSIZE     0.5
    INIT_TYPE           41        # initialization by period relaxation
    DUAL_GAP_TOL        0.01
    SMOOTHING_PARAM     0.1
    PROJ_ROUND_MIDP     0.4
    PERIOD_AGG_POL      52        # aggregating by max (this should be used for UC)
    SCEN_LP_ALG         b         # barrier algorithm
    PERIOD_LP_ALG       b         # barrier algorithm
XPRESS_SCEN_INT
    8278                2         # XPRS_THREADS
    8020                180       # XPRS_MAXTIME
XPRESS_SCEN_DBL
    7020                0.01      # XPRS_MIPRELSTOP
XPRESS_PERIOD_INT
    8278                2         # XPRS_THREADS
    8020                10        # XPRS_MAXTIME
XPRESS_PERIOD_DBL
    7020                0.005     # XPRS_MIPRELSTOP


source/Options.h
/*
Copyright (C) 2020 Ignacio Aravena.

This file is part of AsyncLSD.

AsyncLSD is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

AsyncLSD is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with AsyncLSD. If not, see <https://www.gnu.org/licenses/>.
*/

#ifndef OPTIONS_INCLUDED
	
	// define OPTIONS_INCLUDED to avoid loading the header again
	#define OPTIONS_INCLUDED
	
	/* INCLUDE STATEMENTS */
	
	// Standard libraries
	#include <stdio.h>			// provides functions for reading and printing
	#include <stdlib.h>			// provides several functions for managing memory, files and interact with system
	#include <string.h>			// managing strings
	
	/* CONSTANTS */
	
	// Configuration file constants
	#define CONFIG_NAMECOL			4
	#define CONFIG_VALUECOL			24
	#define CONFIG_MAXLINELEN		1000
	#define CONFIG_MAXNAMELEN		18
	#define CONFIG_MAXNAMELENSTR		"18"
	#define CONFIG_VERBOSITY		"VERBOSITY"
	#define CONFIG_MAXITER			"MAXITER"
	#define CONFIG_MAXTIME			"MAXTIME"
	#define CONFIG_DUAL_SHARE_MIDITER	"DSHARE_MIDITER"		// mid point for dual share change
	#define CONFIG_DUAL_SHARE_START		"DSHARE_START"			// start dual share
	#define CONFIG_DUAL_SHARE_END		"DSHARE_END"			// end dual share
	#define CONFIG_NONSCEN_PERIOD		"NONSCEN_PERIOD"		// evaluation of f0 every how many dual updates (in pu of number of scenarios)
	#define CONFIG_NONSCEN_CENTER		"NONSCEN_CENTER"		// center type for smoothing: fixed center during initializations, moving center with the incumbent or moving center with subgradients (ADMM like)
	#define CONFIG_DUAL_STEPSIZE		"DUAL_STEPSIZE"
	#define CONFIG_STEPSIZE_P		"DUAL_STEPSIZE_P"
	#define CONFIG_STEPSIZE_R		"DUAL_STEPSIZE_R"
	#define CONFIG_STEPSIZE_Q		"DUAL_STEPSIZE_Q"
	#define CONFIG_STEPSIZE_XI		"DUAL_STEPSIZE_XI"
	#define CONFIG_STEPSIZE_SGM		"DUAL_STEPSIZE_SGM"
	#define CONFIG_GNORM_FILTER		"DUAL_GNORM_FILTER"		// Filtering the norm of the subgradient for Polyak stepsie (zero corresponds to Polyak). See Camerini, Fratta, and Maffioli 75.
	#define CONFIG_PRIMAL_RECOVERY		"PRIMAL_RECOVERY"
	#define CONFIG_PRIMAL_IS_SSIZE		"PRIMAL_IS_SSIZE"		// IS sample size for primal recovery
	#define CONFIG_INIT_TYPE		"INIT_TYPE"				// Initialization type: linear relaxation or period relaxation
	#define CONFIG_RELDUALTOL		"DUAL_GAP_TOL"
	#define CONFIG_SMOOTHING_PARAM		"SMOOTHING_PARAM"
	#define CONFIG_PROJ_ROUND_MIDPOINT	"PROJ_ROUND_MIDP"		// mid point for rounding when projecting infeasible first stage solutions (lower is more conservative)
	#define CONFIG_PERIOD_AGG_POLICY	"PERIOD_AGG_POL"		// policy for aggregating "multi-period" first stage variables in period relaxation
	#define CONFIG_SCEN_LP_ALG		"SCEN_LP_ALG"
	#define CONFIG_PERIOD_LP_ALG		"PERIOD_LP_ALG"
	
	// Options constants
	#define VERB_NONE			0
	#define VERB_COORDINATOR_ONLY		1
	#define VERB_COORDINATOR_FUNCS		2
	#define VERB_WORKER			3
	#define VERB_WORKER_FUNCS		4
	#define CENTER_FIXED			10
	#define CENTER_INCUMBENT		11
	#define CENTER_AVERAGE_SUBGRADIENT	12
	#define DUAL_STEPSIZE_CONSTANT		20		// p
	#define DUAL_STEPSIZE_DIMINISHING	21		// p/(1+r/numscenarios*k)^q
	#define DUAL_STEPSIZE_POLYAK		22		// p/(1+r/numscenarios*k)^q*min(xi*max(abs(f(x)),abs(fup)), fup - f(x))/max(sigma, |g|^2_2)
	#define PRIMAL_RECOVERY_FIFO		30		// getting solutions from scenario subproblems as they appear -> requires to keep all solutions
	#define PRIMAL_RECOVERY_RND		31		// getting the solution of one scenario at random from the pile -> requires to keep all solutions
	#define PRIMAL_RECOVERY_LIFO		32		// getting the solution of the lastly solved scenario
	#define PRIMAL_RECOVERY_IS		33		// Importance Sampling recombination heuristic
	#define INIT_TYPE_LINEAR		40		// Use linear relaxation without delayed rows for initialization
	#define INIT_TYPE_PERIOD		41		// Use period relaxation for initialization
	#define PERIOD_POLICY_MIN		50		// Aggregate "multi-period" first stage variables using min 
	#define PERIOD_POLICY_MEAN		51		// Aggregate "multi-period" first stage variables using mean
	#define PERIOD_POLICY_MAX		52		// Aggregate "multi-period" first stage variables using max
	
	/* STRUCTS */
	typedef struct xpress_options {
		char lp_method[2];
		int num_intoptions, num_dbloptions;
		int *intoptions, *dbloptions;
		int *intoptions_value;
		double *dbloptions_value;
	} xpress_options;
	typedef struct options {
		int verbosity, dual_stepsize, primal_recovery;			// caterigorical
		int max_iterations, init_type, period_aggregation_policy;
		double max_time, dual_share_midpoint, dual_share_start, dual_share_end,
			nonscen_evaluation_period, nonscen_center_type;
		double dual_stepsize_p, dual_stepsize_r, dual_stepsize_q, dual_stepsize_xi, dual_stepsize_sigma,
			dual_gnorm2_filter;
		double primal_IS_ssize, rel_dual_gap_tol, dual_f0_mu, proj_round_midpoint;
		xpress_options scenprob_options, periodprob_options;
	} options;
	
	/* FUNCTIONS AND PROCEDURES */
	extern int read_options(const char*, struct options*);
	extern int free_options(struct options*);
	
#endif
